Allow user to choose content warning options when cross-posting from Mastodon to Twitter

singingwolfboy opened the pull request:
As a user crossposting from Mastodon to Twitter, it should be possible to configure how to handle toots with content warnings. Some users may want to display just the content warning, some may want to display just the content, some may want both. The default should be to display just the content warning.

I am not a Ruby on Rails developer, and I probably did this wrong. I know that I'm missing tests, translations, etc, but I wanted to submit what I've managed to put together so far. Hopefully I can get some assistance in getting this feature finished and merged!

renatolond commented:
Hi! Thanks for the pull request!
It's looking good to me so far

Regarding the tests, I usually test only the backend, no need to add tests to the interface.
I can either write them for you or help you write them, what would you prefer? :)

I'll add some comments to the code

renatolond commented on .travis.yml:
```diff
@@ -12,8 +13,10 @@ rvm:
 bundler_args: --without production
 addons:
-    postgresql: "9.6"
+    postgresql: "13"
```
I'd like to keep this compatible with pg down to 9.6 for the moment. The issue, I imagine, is that since the dump is in SQL (rails doesn't support pgsql enums properly), the dump itself is incompatible with 9.6. But I imagine with the changes I added to the dump, it should work with 9.6 too :)

renatolond commented on db/structure.sql:
```diff
@@ -42,6 +29,17 @@ CREATE TYPE public.boost_options AS ENUM (
+-- Name: masto_cw_options; Type: TYPE; Schema: public; Owner: -
+--
+
+CREATE TYPE public.masto_cw_options AS ENUM (
```
If you only keep this block and the last two changes to this file (the one to the users table and the one to the schema migrations), I imagine this will work on 9.6 in travis :)

(In other words, remove the default_with_oids change and three changes before this one)

singingwolfboy replied:
Done! Incidentally, would you like me to submit a pull request to switch from Travis CI to GitHub Actions? [Here's a blog post explaining why that's a good idea.](https://blog.earthly.dev/migrating-from-travis/)

renatolond replied:
Oh, if you have the time I'd love that, I looked quickly into it a few weeks ago but never did the move :) Thanks!

renatolond commented on lib/mastodon_user_processor.rb:
```diff
@@ -282,7 +282,23 @@ def handle_force_url(content)
+          "CW: #{toot.spoiler_text}\n\n#{toot.text_content}"
+        end
+      elsif user.masto_cw_options == "cw_only"
+        "CW: #{toot.spoiler_text} … #{toot.url}"
```
I don't think you need to add the url here, you can use `self.force_toot_url = true`, and it will take care of everything else (including cutting the text if it's too long)

singingwolfboy replied:
I just copied what was in this block originally (line 285 of the file). I'm happy to make that change, but I want to be sure that it's the right thing to do, since it seems unrelated to the core idea of this pull request. (Or do you want to make that change, since it seems like you understand this mechanism better than I do?)

renatolond replied:
Yeah, I think this is legacy code before I introduced the force option, it should behave exactly the same (and in the rare case the CW is too long, it will cut it automagically), I did a quick test on master and it should be fine, you can replace it by:
```
      self.force_toot_url = true
      "CW: #{toot.spoiler_text}"
```

singingwolfboy replied:
Also, I'm very aware that this function has gotten overcomplicated. Maybe it should be refactored to look more like this? (pseudocode)

```
tweet = toot.text_content
has_content = True
if toot is sensitive:
  if user.masto_cw_options == "cw_and_content":
    tweet = "CW: #{toot.spoiler_text}\n\n#{toot.text_content}"
  elsif user.masto_cw_options == "cw_only":
    tweet = "CW: #{toot.spoiler_text}"
    has_content = False
  elsif user.masto_cw_options == "content_only":
    tweet = toot.text_content
  else:
    raise "invalid masto_cw_options"

if should_add_image_count?
  tweet = "#{tweet}... #{toot.media_attachments.count} 🖼️"

return tweet
```

renatolond reviewed (changes requested):
A few changes, but it's looking good overall, thanks again for the PR!

singingwolfboy commented on db/migrate/20210109000000_add_masto_cw_to_twitter_options.rb:
```diff
@@ -0,0 +1,8 @@
+    execute <<-SQL
+      CREATE TYPE masto_cw_options AS ENUM ('CW_AND_CONTENT', 'CONTENT_ONLY', 'CW_ONLY');
+    SQL
+    add_column :users, :masto_cw_options, :masto_cw_options
```
How do I express that the default value should be `CW_ONLY`? As it is, I expect this migration will leave this column at `NULL` by default, and then we need to handle the null case in the code. It would be nice if we could make this column non-nullable, with a default of `CW_ONLY`, so that we don't need to handle the null case in the application code.

renatolond replied:
You can use it as if it were a string, like so:
`add_column :users, :masto_cw_options, :masto_cw_options, default: 'CW_ONLY'`

singingwolfboy commented:
Codeclimate is unhappy with the complexity of the method, but I don't see a way to make it simpler. Any suggestions?

singingwolfboy commented:
> Regarding the tests, I usually test only the backend, no need to add tests to the interface.
> I can either write them for you or help you write them, what would you prefer? :)

I took a look at the test file, and unfortunately I don't really understand what's going on there. If you could write the tests, I would appreciate it.

renatolond commented:
> Codeclimate is unhappy with the complexity of the method, but I don't see a way to make it simpler. Any suggestions?

Don't worry about it, I don't use cyclomatic usually, I think in the last rubocop upgrade I forgot to disable it again :)

renatolond commented:
I pushed some tests and I also updated the main branch to disable the metrics in rubocop, so if you rebase you codeclimate should go away :)
The only thing left to do is to add translations for en: you can edit `config/locales/en.yml`, and you need to add the new keys you just added: `masto_cw_options`, `masto_cw_options_explanation` and one for each of the new options you added `cw_only`, `content_only` and `cw_and_content` :)
Thanks again!

renatolond commented:
Fixes #137

singingwolfboy commented:
Alright, I've added the translations and rebased! I think these commits should probably be squashed together when merged, since the commit history doesn't really add anything meaningful, but I leave that up to you.

Thanks for your help, @renatolond! My first PR to a Ruby on Rails project 🎉

singingwolfboy commented:
Thanks for merging this -- and it looks like you've deployed it already! Thanks!

However, I've already found a bug. When I select one of the other options aside from the default, and then click the "Update User" button, the banner at the top of the page says "Your changes were saved" but the radio button is back at the default value of "CW only". I don't know if it's getting saved correctly and the UI is wrong, or if it's not getting saved at all. What do you think?

renatolond commented:
Ah, indeed! Sorry, my bad, I forgot one small piece: in the controller for the options, `app/controllers/users_controller.rb`, [rails has a list of the params that it permits](https://edgeapi.rubyonrails.org/classes/ActionController/StrongParameters.html). Anything out of those is ignored to avoid assignment of params you do not allow the user to access. You need to add the new param, `masto_cw_options` to that list of permitted params and it should work :)
